File: src/main/java/no/novari/flyt/egrunnerverv/gateway/dispatch/converting/JournalpostToInstanceReceiptDispatchEntityConvertingService.java
package no.novari.flyt.egrunnerverv.gateway.dispatch.converting;

import no.fint.model.felles.basisklasser.Begrep;
import no.fint.model.felles.kompleksedatatyper.Identifikator;
import no.fint.model.resource.administrasjon.personal.PersonalressursResource;
import no.fint.model.resource.arkiv.kodeverk.JournalStatusResource;
import no.fint.model.resource.arkiv.kodeverk.JournalpostTypeResource;
import no.fint.model.resource.arkiv.kodeverk.SkjermingshjemmelResource;
import no.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource;
import no.fint.model.resource.arkiv.noark.AdministrativEnhetResource;
import no.fint.model.resource.arkiv.noark.ArkivressursResource;
import no.fint.model.resource.arkiv.noark.JournalpostResource;
import no.fint.model.resource.arkiv.noark.SakResource;
import no.fint.model.resource.arkiv.noark.SkjermingResource;
import no.fint.model.resource.felles.PersonResource;
import no.novari.cache.FintCache;
import no.novari.flyt.egrunnerverv.gateway.dispatch.model.JournalpostReceipt;
import org.springframework.stereotype.Service;

import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static no.novari.flyt.egrunnerverv.gateway.ResourceLinkUtil.getOptionalFirstLink;
import static no.novari.flyt.egrunnerverv.gateway.dispatch.converting.InstanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService.EGRUNNERVERV_DATETIME_FORMAT;

@Service
public class JournalpostToInstanceReceiptDispatchEntityConvertingService {

    private final FintCache<String, AdministrativEnhetResource> administrativEnhetResourceCache;
    private final FintCache<String, ArkivressursResource> arkivressursResourceCache;
    private final FintCache<String, JournalStatusResource> journalStatusResourceCache;
    private final FintCache<String, JournalpostTypeResource> journalpostTypeResourceCache;
    private final FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache;
    private final FintCache<String, SkjermingshjemmelResource> skjermingshjemmelResourceCache;
    private final FintCache<String, PersonalressursResource> personalressursResourceCache;
    private final FintCache<String, PersonResource> personResourceCache;

    public JournalpostToInstanceReceiptDispatchEntityConvertingService(
            FintCache<String, AdministrativEnhetResource> administrativEnhetResourceCache,
            FintCache<String, ArkivressursResource> arkivressursResourceCache,
            FintCache<String, JournalStatusResource> journalStatusResourceCache,
            FintCache<String, JournalpostTypeResource> journalpostTypeResourceCache,
            FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache,
            FintCache<String, SkjermingshjemmelResource> skjermingshjemmelResourceCache,
            FintCache<String, PersonalressursResource> personalressursResourceCache,
            FintCache<String, PersonResource> personResourceCache
    ) {
        this.administrativEnhetResourceCache = administrativEnhetResourceCache;
        this.arkivressursResourceCache = arkivressursResourceCache;
        this.journalStatusResourceCache = journalStatusResourceCache;
        this.journalpostTypeResourceCache = journalpostTypeResourceCache;
        this.tilgangsrestriksjonResourceCache = tilgangsrestriksjonResourceCache;
        this.skjermingshjemmelResourceCache = skjermingshjemmelResourceCache;
        this.personalressursResourceCache = personalressursResourceCache;
        this.personResourceCache = personResourceCache;
    }

    public JournalpostReceipt map(SakResource sakResource, Long journalpostNummer) {
        Optional<PersonalressursResource> saksansvarligPersonalressursResource =
                getOptionalFirstLink(sakResource::getSaksansvarlig)
                        .flatMap(arkivressursResourceCache::getOptional)
                        .flatMap(arkivressurs -> getOptionalFirstLink(arkivressurs::getPersonalressurs))
                        .flatMap(personalressursResourceCache::getOptional);

        Optional<PersonResource> saksansvarligPersonResource =
                saksansvarligPersonalressursResource
                        .flatMap(personalressursResource -> getOptionalFirstLink(personalressursResource::getPerson))
                        .flatMap(personResourceCache::getOptional);

        JournalpostResource journalpostResource = sakResource.getJournalpost()
                .stream()
                .filter(journalpost -> Objects.equals(journalpost.getJournalPostnummer(), journalpostNummer))
                .findFirst()
                .orElseThrow(
                        () -> new IllegalArgumentException("No journalpost with journalpostNummer=" + journalpostNummer)
                );

        Optional<AdministrativEnhetResource> administrativEnhetResource =
                getOptionalFirstLink(journalpostResource::getAdministrativEnhet)
                        .flatMap(administrativEnhetResourceCache::getOptional);

        Optional<JournalStatusResource> journalStatusResource =
                getOptionalFirstLink(journalpostResource::getJournalstatus)
                        .flatMap(journalStatusResourceCache::getOptional);

        Optional<JournalpostTypeResource> journalpostTypeResource =
                getOptionalFirstLink(journalpostResource::getJournalposttype)
                        .flatMap(journalpostTypeResourceCache::getOptional);

        Optional<SkjermingResource> skjermingResource = Optional.ofNullable(journalpostResource.getSkjerming());

        Optional<TilgangsrestriksjonResource> tilgangsrestriksjonResource =
                skjermingResource.flatMap(sr -> getOptionalFirstLink(sr::getTilgangsrestriksjon))
                        .flatMap(tilgangsrestriksjonResourceCache::getOptional);

        Optional<SkjermingshjemmelResource> skjermingshjemmelResource =
                skjermingResource.flatMap(sr -> getOptionalFirstLink(sr::getSkjermingshjemmel))
                        .flatMap(skjermingshjemmelResourceCache::getOptional);

        JournalpostReceipt.JournalpostReceiptBuilder builder =
                JournalpostReceipt
                        .builder()
                        .journalpostnr("%s-%s".formatted(
                                sakResource.getMappeId().getIdentifikatorverdi(),
                                journalpostResource.getJournalPostnummer()
                        ))
                        .tittel(journalpostResource.getTittel())
                        .dokumentdato(formatDate(journalpostResource.getOpprettetDato()));

        journalStatusResource
                .map(Begrep::getSystemId)
                .map(Identifikator::getIdentifikatorverdi)
                .ifPresent(builder::statusId);

        tilgangsrestriksjonResource
                .map(Begrep::getKode)
                .ifPresent(builder::tilgangskode);

        skjermingshjemmelResource
                .map(Begrep::getSystemId)
                .map(Identifikator::getIdentifikatorverdi)
                .ifPresent(builder::hjemmel);

        journalpostTypeResource
                .map(Begrep::getSystemId)
                .map(Identifikator::getIdentifikatorverdi)
                .ifPresent(builder::dokumenttypeid);

        saksansvarligPersonalressursResource
                .map(PersonalressursResource::getBrukernavn)
                .map(Identifikator::getIdentifikatorverdi)
                .ifPresent(builder::saksansvarligbrukernavn);

        saksansvarligPersonResource.map(this::fullName)
                .ifPresent(builder::saksansvarlignavn);

        administrativEnhetResource
                .map(AdministrativEnhetResource::getSystemId)
                .map(Identifikator::getIdentifikatorverdi)
                .ifPresent(builder::adminenhetkortnavn);

        administrativEnhetResource
                .map(AdministrativEnhetResource::getNavn)
                .ifPresent(builder::adminenhetnavn);

        return builder.build();
    }

    private String formatDate(Date date) {
        if (date == null) return null;
        return date.toInstant()
                .atZone(ZoneId.systemDefault())
                .toLocalDateTime()
                .format(DateTimeFormatter.ofPattern(EGRUNNERVERV_DATETIME_FORMAT));
    }

    private String fullName(PersonResource person) {
        return Stream.of(
                        person.getNavn().getFornavn(),
                        person.getNavn().getMellomnavn(),
                        person.getNavn().getEtternavn()
                ).filter(Objects::nonNull)
                .collect(Collectors.joining(" "));
    }

}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/dispatch/converting/JournalpostToInstanceReceiptDispatchEntityConvertingService.kt
package no.novari.flyt.egrunnerverv.gateway.dispatch.converting

import no.fint.model.resource.administrasjon.personal.PersonalressursResource
import no.fint.model.resource.arkiv.kodeverk.JournalStatusResource
import no.fint.model.resource.arkiv.kodeverk.JournalpostTypeResource
import no.fint.model.resource.arkiv.kodeverk.SkjermingshjemmelResource
import no.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource
import no.fint.model.resource.arkiv.noark.AdministrativEnhetResource
import no.fint.model.resource.arkiv.noark.ArkivressursResource
import no.fint.model.resource.arkiv.noark.JournalpostResource
import no.fint.model.resource.arkiv.noark.SakResource
import no.fint.model.resource.arkiv.noark.SkjermingResource
import no.fint.model.resource.felles.PersonResource
import no.novari.cache.FintCache
import no.novari.flyt.egrunnerverv.gateway.ResourceLinkUtil.getOptionalFirstLink
import no.novari.flyt.egrunnerverv.gateway.dispatch.model.JournalpostReceipt
import org.springframework.stereotype.Service
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Date
import java.util.Objects
import java.util.stream.Collectors
import java.util.stream.Stream

@Service
class JournalpostToInstanceReceiptDispatchEntityConvertingService(
    private val administrativEnhetResourceCache: FintCache<String, AdministrativEnhetResource>,
    private val arkivressursResourceCache: FintCache<String, ArkivressursResource>,
    private val journalStatusResourceCache: FintCache<String, JournalStatusResource>,
    private val journalpostTypeResourceCache: FintCache<String, JournalpostTypeResource>,
    private val tilgangsrestriksjonResourceCache: FintCache<String, TilgangsrestriksjonResource>,
    private val skjermingshjemmelResourceCache: FintCache<String, SkjermingshjemmelResource>,
    private val personalressursResourceCache: FintCache<String, PersonalressursResource>,
    private val personResourceCache: FintCache<String, PersonResource>,
) {
    fun map(
        sakResource: SakResource,
        journalpostNummer: Long,
    ): JournalpostReceipt {
        val saksansvarligPersonalressursResource =
            getOptionalFirstLink(sakResource::getSaksansvarlig)
                ?.let(arkivressursResourceCache::getOptional)
                ?.orElse(null)
                ?.let { arkivressurs ->
                    getOptionalFirstLink(arkivressurs::getPersonalressurs)
                        ?.let(personalressursResourceCache::getOptional)
                        ?.orElse(null)
                }

        val saksansvarligPersonResource =
            saksansvarligPersonalressursResource
                ?.let { personalressursResource ->
                    getOptionalFirstLink(personalressursResource::getPerson)
                        ?.let(personResourceCache::getOptional)
                        ?.orElse(null)
                }

        val journalpostResource =
            sakResource.journalpost
                .stream()
                .filter { journalpost: JournalpostResource ->
                    Objects.equals(journalpost.journalPostnummer, journalpostNummer)
                }.findFirst()
                .orElseThrow {
                    IllegalArgumentException("No journalpost with journalpostNummer=$journalpostNummer")
                }

        val administrativEnhetResource =
            getOptionalFirstLink(journalpostResource::getAdministrativEnhet)
                ?.let(administrativEnhetResourceCache::getOptional)
                ?.orElse(null)

        val journalStatusResource =
            getOptionalFirstLink(journalpostResource::getJournalstatus)
                ?.let(journalStatusResourceCache::getOptional)
                ?.orElse(null)

        val journalpostTypeResource =
            getOptionalFirstLink(journalpostResource::getJournalposttype)
                ?.let(journalpostTypeResourceCache::getOptional)
                ?.orElse(null)

        val skjermingResource: SkjermingResource? = journalpostResource.skjerming

        val tilgangsrestriksjonResource =
            skjermingResource
                ?.let { sr ->
                    getOptionalFirstLink(sr::getTilgangsrestriksjon)
                        ?.let(tilgangsrestriksjonResourceCache::getOptional)
                        ?.orElse(null)
                }

        val skjermingshjemmelResource =
            skjermingResource
                ?.let { sr ->
                    getOptionalFirstLink(sr::getSkjermingshjemmel)
                        ?.let(skjermingshjemmelResourceCache::getOptional)
                        ?.orElse(null)
                }

        var receipt =
            JournalpostReceipt(
                journalpostnr = "${sakResource.mappeId.identifikatorverdi}-${journalpostResource.journalPostnummer}",
                tittel = journalpostResource.tittel,
                statusId = null,
                tilgangskode = null,
                hjemmel = null,
                dokumentdato = formatDate(journalpostResource.opprettetDato),
                dokumenttypeid = null,
                dokumenttypenavn = null,
                saksansvarligbrukernavn = null,
                saksansvarlignavn = null,
                adminenhetkortnavn = null,
                adminenhetnavn = null,
            )

        journalStatusResource
            ?.systemId
            ?.identifikatorverdi
            ?.let { statusId -> receipt = receipt.copy(statusId = statusId) }

        tilgangsrestriksjonResource
            ?.kode
            ?.let { tilgangskode -> receipt = receipt.copy(tilgangskode = tilgangskode) }

        skjermingshjemmelResource
            ?.systemId
            ?.identifikatorverdi
            ?.let { hjemmel -> receipt = receipt.copy(hjemmel = hjemmel) }

        journalpostTypeResource
            ?.systemId
            ?.identifikatorverdi
            ?.let { dokumenttypeid -> receipt = receipt.copy(dokumenttypeid = dokumenttypeid) }

        saksansvarligPersonalressursResource
            ?.brukernavn
            ?.identifikatorverdi
            ?.let { brukernavn -> receipt = receipt.copy(saksansvarligbrukernavn = brukernavn) }

        saksansvarligPersonResource
            ?.let(this::fullName)
            ?.let { navn -> receipt = receipt.copy(saksansvarlignavn = navn) }

        administrativEnhetResource
            ?.systemId
            ?.identifikatorverdi
            ?.let { kortnavn -> receipt = receipt.copy(adminenhetkortnavn = kortnavn) }

        administrativEnhetResource
            ?.navn
            ?.let { navn -> receipt = receipt.copy(adminenhetnavn = navn) }

        return receipt
    }

    private fun formatDate(date: Date?): String? =
        date
            ?.toInstant()
            ?.atZone(ZoneId.systemDefault())
            ?.toLocalDateTime()
            ?.format(DateTimeFormatter.ofPattern(EGRUNNERVERV_DATETIME_FORMAT))

    private fun fullName(person: PersonResource): String =
        Stream
            .of(
                person.navn.fornavn,
                person.navn.mellomnavn,
                person.navn.etternavn,
            ).filter(Objects::nonNull)
            .collect(Collectors.joining(" "))

    companion object {
        private const val EGRUNNERVERV_DATETIME_FORMAT =
            InstanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService.EGRUNNERVERV_DATETIME_FORMAT
    }
}
